zebrad/src/components/inbound.rs

```diff
@@ -178,6 +192,14 @@ impl Service<zn::Request> for Inbound {
                 None => async { Err("not ready to serve addresses".into()) }.boxed(),
             },
             zn::Request::BlocksByHash(hashes) => {
+                // Correctness:
+                //
+                // We don't need to use ServiceExt::oneshot here, because
+                // call_all uses poll_ready internally.
+                //
+                // The state must not return Poll::Pending, because call_all
```

Comment:
I don't think we can guarantee this invariant, because the state is wrapped in a buffer.

Reply:
ready! returns Pending when the buffer is full:
https://github.com/tower-rs/tower/blob/master/tower/src/buffer/service.rs#L119
https://github.com/tower-rs/tower/blob/master/tower/src/semaphore.rs#L55

So we should stop using call_all, because buffers can always return Poll:Pending, and that triggers a bug in call_all.

Instead, we can just use a loop inside the returned future. Correctness and simplicity is more important that a possible small performance gain from doing concurrent block queries for the same peer request.

Reply:
I had to stop using future combinators due to ownership and type issues, so this code is now an async block with a loop.